04-project specific filtering and run script

kgaonkar6 opened the pull request:
<!--Hi there, thanks for your contribution! Please take a moment to fill out this template to facilitate the review of your pull request.-->


#### Purpose/implementation
Project specific filtering to gather recurrent fusions and remove artifacts found in multiple broad histologies.
Putative driver fusions using Gene[1A|2A|1B|2B]_anno


_What scientific question is your analysis addressing? 
What was your approach?
- [X] In-frame fusions is called by both callers
- [X] In-frame fusions is called in atleast 2 samples per histology
- [X] Fusions with either gene fused more than 5 times in a histology
- [X] Fusions in Filtered-fusions found in more than 1 histology

If this is not adding an analysis, describe your changes in this section._

#### Issue

_What GitHub issue does your pull request address?_

#### Directions for reviewers

_Tell potential reviewers what kind of feedback you are soliciting. 
Are there particular areas that need a closer look? Add/Edit project based project specific filtering fusion . 
  
Is there something you want to discuss further? Any stats required for summary?
Should the filtering criteria be annotated in the final files or removed? 
Right now the fusions that don't pass the filters are being removed to make the final FIlteredFusion.tsv and PutativeDriverFusion.tsv

Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review? 
Yes FilteredFusion.tsv and PuativeDriverFusion.tsv tables in results/


#### Results

_If your pull request includes code that produces scientific results, please summarize the results here.

Filtered fusions have fusions which are recurrent and artifact QC-ed results
Putative Driver fusions which are cancer specific fusions

This can help facilitate discussion around interpretation.
Please state what kinds of results are included (e.g., table, figure): table


#### Docker and continuous integration

_Check all those that apply or remove this section if it is not applicable._

- [X] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [ ] This analysis has been added to continuous integration.

Parts to this analysis
- [x] Fusion standardization
- [x] Fusion filtering (general)
- [x] Putative driver fusion expression filtering 
- [x] Filtering fusion expression filtering
- [x] GeneFusion Annotation 
- [x] Gtex/cohort zscore annotation
- [x] Project specific filtering

kgaonkar6 commented:
Hi @jaclyn-taroni I pushed a run_fusion_filtering.sh to run the 01-03 parts for starfusion and arriba using polya and stranded and it looks like it runs correctly locally for the whole set of samples.

But there is an error while reading a file in 04-project-specific-filtering.Rmd , would that be because of the smaller set of samples in the circleci or is there an issue with the location of some file here?

jaclyn-taroni commented:
Hm, it is a bit difficult to tell what's going on in CI from what is here: https://circleci.com/gh/AlexsLemonade/OpenPBTA-analysis/730?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

I would add a commit with the following changes:
- Comment out the project specific filtering step in the config file (the one with `./scripts/run_in_ci.sh Rscript -e "rmarkdown::render('analyses/fusion_filtering/04-project-specific-filtering.Rmd')"`)
- Add a step in the config file that lists the `scratch` directory contents.

I would also add
```
set -e
set -o pipefail
```
to the top of your bash script. If you want to see an example of why that's helpful in CI and a little more background, you can see me fixing my mistake in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/141 😄

jaclyn-taroni commented:
I'm going to remove the quotes for the `ls scratch` step to see if we can get more information.

jaclyn-taroni commented:
Looks like this is what is in `scratch`:
```
arribaPolya_QC_expression_filtered_annotated.RDS
arribaPolya_QC_expression_GTExComparison_annotated.RDS
arribaPolya_unfiltered_annotated.RDS
arribaStranded_QC_expression_filtered_annotated.RDS
arribaStranded_QC_expression_GTExComparison_annotated.RDS
arribaStranded_unfiltered_annotated.RDS
arriba.tsv
metadata_filtered_maf_samples.tsv
mutect2.RDS
starfusionPolya_QC_expression_filtered_annotated.RDS
starfusionPolya_QC_expression_GTExComparison_annotated.RDS
starfusionPolya_unfiltered_annotated.RDS
starfusionStranded_QC_expression_filtered_annotated.RDS
starfusionStranded_unfiltered_annotated.RDS
starfusion.tsv
strafusionStranded_QC_expression_GTExComparison_annotated.RDS
strelka2.RDS
```
I would double check your filenames in the notebook. `scratch/starfusionPolya_QC_expression_filtered_GTExComparison_annotated.RDS` doesn't appear to be in that list: https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/166/files#diff-529ecb14fbf075deaea62321c4eea8f9R23

kgaonkar6 commented:
Yup there was a typo and names in run script were not right :/ sorry about that!

> I would double check your filenames in the notebook. `scratch/starfusionPolya_QC_expression_filtered_GTExComparison_annotated.RDS` doesn't appear to be in that list: https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/166/files#diff-529ecb14fbf075deaea62321c4eea8f9R23

kgaonkar6 commented on analyses/fusion_filtering/04-project-specific-filtering.Rmd:
```diff
@@ -0,0 +1,221 @@
+  dplyr::mutate(Caller = toString(Caller), caller.count = n()) %>%
+  dplyr::filter(caller.count >= numCaller) %>%
+  unique() %>%
+  mutate(note=paste0("Called by",numCaller, "callers")) %>%
```
not really I don't use that column anywhere later, I just add them there if we need some annotation for why they were removed.

kgaonkar6 commented on analyses/fusion_filtering/04-project-specific-filtering.Rmd:
```diff
@@ -0,0 +1,221 @@
+---
+title: "Project specific filtering"
+output: html_notebook
+params:
```
Wooaah!! Whenever I do something stupid in R (which is very often) I'm going to look up this comment up for a confidence boost!  :D

kgaonkar6 commented on analyses/fusion_filtering/run_fusion_filtering.sh:
```diff
@@ -0,0 +1,31 @@
+
+set -e
+set -o pipefail
+
```
Yeah good idea to set up variables for files and and long format param values!! Definitely will help in giving the correct values. :/

jaclyn-taroni replied:
I've found it's often about making it harder to do something you did not intend to do or harder give the incorrect values because there are of course limits to human attention and memory 🙂

kgaonkar6 commented:
I was doing some checks mainly looking at recurrent fusion calls between the previous outputs results where we used a combined starfusion + arriba standard fusion call set as input within the scripts, we were now running starfusion and arriba in separate runs with polya and stranded expression files and found that these 2 edits needed to me made :

1 . Readthroughs that are only annotated by arriba were removed from arriba calls but kept in starfusion calls in the final results since STARfusion doesn't have them annotated as readthroughs by FusionAnnotator. To implement this filter correctly I'm saving all the readthroughs to a file without the distance and removing them in project specific filtering step which I have added here: 
https://github.com/kgaonkar6/OpenPBTA-analysis/blob/f1431402b4866381a6578ce8c82f1a1f43c8e939/analyses/fusion_filtering/02-fusion-filtering.R#L123

Should I read the starfusion and arriba fusion files and merge them so that we can run 02 and 03 on a merged call set, would that be ok?

2. PutativeDriver and FilteredFusion had an additional filter to remove multi-fused genes and read throughs which I have added here: 
https://github.com/kgaonkar6/OpenPBTA-analysis/blob/f1431402b4866381a6578ce8c82f1a1f43c8e939/analyses/fusion_filtering/04-project-specific-filtering.Rmd#L240

jharenza commented on Dockerfile:
```diff
@@ -78,5 +78,10 @@ RUN apt-get update -qq && apt-get -y --no-install-recommends install \
     glmnetUtils
 
+# Required for easy easy regex functions
+RUN R -e "devtools::install_github('trinker/qdapRegex',dependencies = TRUE)"
```
should be filtering all readthrough, agnostic of algorithm for this

kgaonkar6 replied:
not really I'll update that to install.packages from CRAN

kgaonkar6 commented on analyses/fusion_filtering/04-project-specific-filtering.Rmd:
```diff
@@ -0,0 +1,262 @@
+limitMultiFused<-params$limitMultiFused
+
+# remove read-throughs from all callers
+readthrough<-read_tsv("scratch/readthroughs.tsv")
```
I removed that now that I'm merging the callers upstream in 02-fusion-filtering

kgaonkar6 commented on analyses/fusion_filtering/04-project-specific-filtering.Rmd:
```diff
@@ -0,0 +1,262 @@
+  dplyr::select(-JunctionReadCount,-SpanningFragCount) %>%
+  group_by(FusionName, Sample ) %>%
+  unique() %>%
+  dplyr::mutate(Caller = toString(Caller), caller.count = n())
```
Ah no.. that is a problem . I only need to capture calls that are in bother callers

kgaonkar6 commented on analyses/fusion_filtering/04-project-specific-filtering.Rmd:
```diff
@@ -0,0 +1,262 @@
+suppressPackageStartupMessages(library("qdapRegex"))
+
+#read filtFusion files
+arribaStrandedQCGeneFiltered_filtFusion<-readRDS(params$dataStrandedARRIBA)
```
yes that's right

kgaonkar6 commented on analyses/fusion_filtering/04-project-specific-filtering.Rmd:
```diff
@@ -0,0 +1,262 @@
+  dplyr::distinct() %>%
+  group_by(Sample,GeneSymbol) %>% 
+  dplyr::summarise(Gene.ct = n()) %>% 
+  dplyr::filter(Gene.ct>limitMultiFused) %>%
```
yes it's to remove these genes from putative driver list

jaclyn-taroni commented:
@kgaonkar6 I am having a bit of trouble keeping up with the changes in this pull request because there have been multiple rounds of changes to multiple files. Would it be possible to split this into 3 pull requests:

1. Upstream changes to `02`, `03` and the `Dockerfile` because that includes a dependency required for changes in `02`.
2. The shell script for running the merged pipeline (I don't know that you need the other one anymore).
3. `04-project-specific-filtering.Rmd`

A big reason why I ask folks to split things up is because as pull requests grow, the quality of the code review decreases. I don't want to miss something in the upstream changes because the GitHub diff hides that file and I inadvertently skip it.

@cansavvy has been splitting quite a few pull requests up lately (see: https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/159#issuecomment-546332184 and https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/174#issuecomment-547839034). I'm sure she'd be happy to help you out over in the `#open-pbta` channel in Slack if you need help.

kgaonkar6 commented:
Sorry about that!  I was editing the upstream files for merging the 2 callers and implementing suggestions from you for project specific. I can close this and get help from @cansavvy for the new PRs. Thanks!
